lib/Transforms/Functional.cpp

```diff
@@ -0,0 +1,153 @@
+//===- Functional.cpp -------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Transforms/Functional.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+
+using namespace mlir;
+using functional::impl::SequencePattern;
+
+Operation *mlir::functional::findNearestScope(Operation *op) {
+  do {
+    op = op->getParentOp();
+  } while (op && !op->hasTrait<OpTrait::IsIsolatedFromAbove>());
+  return op;
+}
+
+SmallVector<std::unique_ptr<Region>>
+mlir::functional::cloneRegionsOf(Operation *op) {
+  SmallVector<std::unique_ptr<Region>> regions;
+  BlockAndValueMapping mapping;
+  for (Region &region : op->getRegions()) {
+    region.cloneInto(&*regions.emplace_back(std::make_unique<Region>()),
```

Comment:
nit: maybe if you s/region/r and s/mapping/bvm you can save 1 line + braces (i.e. 2 lines total)?